Adding Black Magic Design Decklink support to FFMPEG Preset

masterav10 opened the pull request:
This pull request enables decklink device support if the appropriate Black Magic SDK is installed on the build machine.

Currently enabled for only Windows x86_64. The SDK supports windows, linux, and mac, so we can probably add decklink to those platforms as well.

I have verified that the devices are enumerated through libavdevice on a machine with a Black Magic Decklink Quad. Haven't had a chance to hook up to the them, though.

In addition to this pull request, we need to do the following:
1. Update msys config to include the following:
`pacman -S mingw-w64-{i686,x86_64}-tools-git dos2unix`
2. Install Black Magic Desktop SDK on the appveyor server.

If we don't want the backoff approach as taken here, we could include the headers directly into the repository under src/main/resources. Then we don't have to the above. However, it places more burden on the maintainer, since you have to download the SDK and run midl to get the appropriate headers. Let me know if you'd rather change to this approach.

masterav10 committed:
Adding support for Decklink Devices from Black Magic Design.

This commit adds support for windows-x86_64, but plan to add support for all platforms.

masterav10 committed:
Enabling decklink support.

masterav10 committed:
Attempting to create static linking for avdevice.

masterav10 committed:
Merge branch 'master' into ffmpeg-decklink

masterav10 committed:
An additional patch to static link to libstdc++.

Also need to install this for widl.
pacman -S mingw-w64-{i686,x86_64}-tools-git

masterav10 committed:
Decklink support for windows x86-64 if SDK is installed.

masterav10 committed:
Moved patching outside of conditional to ensure paths are correct.

masterav10 committed:
Merge branch 'master' into ffmpeg-decklink

saudet commented:
This links against a dynamic library we can't bundle, am I correct?

masterav10 commented:
Shouldn't be. Which library are you referring to? The only thing it should be linked to are native libraries. Here is the one I built yesterday before the pull request.

![image](https://user-images.githubusercontent.com/3239352/38170488-b3182686-3554-11e8-99ea-adcbd0c2ab29.png)

saudet commented:
Ok, looks good. Please add the installation of the SDK in the `install-windows.sh` script, and do the same for Linux and Mac, if possible. Thanks!

masterav10 commented:
Yeah, I'm going to have to make some changes to make this work. There is no way to download the SDK automatically that I can see (used to be available). And even if we could, there license for the SDK as a whole is different than the license for the individual files (which is more permissive).

Next best option is to simply copy the appropriate header files directly into the repo. When a new update occurs, simple download the new SDK and install the appropriate headers again. Annoying, but doable.

[Zeronae used to support it but removed said support, claiming incompatible license.](https://ffmpeg.zeranoe.com/forum/viewtopic.php?f=42&t=1823&sid=84314c1026aecc61aeec18fb9f096d55&start=50#p12540)

I cannot find any section in either of the license agreements that would cause a conflict with an existing license we have, and as you have seen, we don't like to any binaries. So I'm going to proceed forth with the intent of adding decklink support. I'll ping when I'm done.

saudet commented:
If FFmpeg just needs a few small header files, we can just add them to this repository in `src/main/resources/` and use them from there.

masterav10 committed:
Adding all of the decklink headers directly, based on licensing issues.

masterav10 committed:
Typo.

masterav10 commented:
The following platforms should now support decklink via FFMPEG:
macosx, linux-x86, linux-x86_64, windows-x86, windows-x86_64.

Verified the Windows build with my local machine. Will test against an actual decklink in the coming week.

saudet commented:
Wow, that's a lot of header files, let's put that somewhere else...

masterav10 commented:
No problem. Will make a separate repo here on github. Gonna be a couple of days before I get to it, potentially.

masterav10 commented:
That's for the drivers. We need the SDK. Or did you find a different link?

saudet commented:
I see. These files already appear to be available in other repositories though, for example:
https://github.com/GStreamer/gst-plugins-bad/tree/master/sys/decklink/linux
We could just pull from there maybe? Although the license does appear to be a concern:
https://patchwork.ffmpeg.org/patch/3230/
We might want to contact DeckLink directly to sort this out...

masterav10 commented:
Looking into this.

masterav10 committed:
Fixing a bug with the decklink implementation.

I can now detect when a source is not available using the flags field of the AVPacket.

masterav10 commented:
Finally got an answer back. The EULA for the SDK is designed so that a single public point is available for downloading the SDK. This does in fact prevent us from using their headers, since the GPL requires that we provide all sources (including the decklink headers) to anyone we distribute our libraries to.

If possible, I would like to revert back to my previous implementation which will detect the SDK on the machine where the build is happening. Since we won't have the SDK installed on the CI servers, it won't add decklink to your normal distribution. Then I can make local builds of FFmpeg with decklink support without the need to maintain a separate branch.

Seem reasonable?

saudet commented:
We don't need to redistribute it under GPL, but it can't be redistributed
at all. So I guess we'd have to do like with FlyCapture and try to make the
SDK only accessible to the CI servers. If you're satisfied with local
builds for now though, that's fine, yes.

masterav10 committed:
Prevent logging from generation too many threads.

masterav10 committed:
Removed unredistributable files.

Added a patch to remove references to unredistributable files from within the FFmpeg source code.

masterav10 committed:
Removed unredistributable files.

Added a patch to remove references to unredistributable files from within the FFmpeg source code.

masterav10 committed:
Upgraded patches to support new ffmpeg version.

saudet commented:
For now though, please remove from this PR the files from the SDK so we can merge this! Thanks

masterav10 commented:
I should be able to finish this by the weekend.

masterav10 committed:
Merge branch 'master' of https://github.com/bytedeco/javacpp-presets into ffmpeg-decklink

masterav10 committed:
Detached thread was for testing, removing.

masterav10 commented:
This adds support for windows 32 and 64. I've only tested the 64-bit version, but it works.

Added some patches to ignore SDK files; this keeps our license kosher. There is also a patch to the FFmpeg decklink code so that I can tell if a video source stops producing frames. Let me know if you need anything else.

saudet commented:
Looking good, but DeckLinkAPI.h and DeckLinkAPI_i.c are still there, could you remove them? Let's also remove LICENSE.DeckLinkAPI since it is misleading.

Could we also consolidate all the patches into one big decklink.patch? Thanks!

saudet commented:
Are you saying that DeckLinkAPI.h and DeckLinkAPI_i.c are not part of the SDK? Sounds like they were generated from the SDK... And what's with the DeckLinkAPI.LICENSE file?

masterav10 commented:
They are indeed generated from the SDK. I believe that allows them to fall under the permissive license (which I included). Would you prefer them elsewhere? I guess we could generate them, but you'd have to install the SDK on the server, along with some additional msys libs.

I will take care of the patches.

saudet commented:
Right, so maybe it's alright. Is this what your contact at DeckLink said?

masterav10 commented:
That is correct; the files in the SDK cannot be distributed, but files generated from the SDK can be distributed under the permissive license.

saudet commented:
Is this in the license that came with the SDK you have?
> Developer agrees not to distribute the Materials or **any derivative works** created therewith without the express written permission of an authorized Blackmagic Design officer or employee.

https://notabug.org/RiCON/decklink-headers/raw/d0ead620884ab95954f55646f1acddb100b7ebdf/SDK%20License.txt

It sounds like we would still need their permission. Can you forward me the message you got from DeckLink about this?

masterav10 commented:
That is not the correct license. You can read the full license by attempting to download the SDK (copy it into notepad). Derivative works are not mentioned.

masterav10 committed:
Combining patches using combinediff.

masterav10 committed:
Did not commit the patch.

saudet commented:
@masterav10 Yeah, if we could have them look at this PR just to make sure that would be great. Otherwise looks good, though does this not build on Linux or Mac? We'd need something more for that?

saudet commented:
Any news from them? Have the OKed this PR?

masterav10 commented:
No word yet. I could try to send another email.
